In [1]:
import pandas as pd
data=pd.read_csv("C:/Users/sharv/Documents/Post Graduate Program In Data Science-PurdueSimpliLearn/Course3-Machine Learning/Projects/Amazon - Movies and TV Ratings.csv")
data.tail()

user_id  Movie1  Movie2  Movie3  Movie4  Movie5  Movie6  Movie7  \
4843  A1IMQ9WMFYKWH5     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4844  A1KLIKPUF5E88I     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4845   A5HG6WFZLO10D     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4846  A3UU690TWXCG1X     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4847   AI4J762YI6S06     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

      Movie8  Movie9  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
4843     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4844     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4845     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4846     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   
4847     NaN     NaN  ...       NaN       NaN       NaN       NaN       NaN   

      Movie202  Movie203  Movie204  Movie205  Movie206  
4843       NaN       NaN       NaN       NaN       5.0  
4844       NaN       NaN       NaN       NaN       5.0  
4845       NaN       NaN       NaN       NaN       5.0  
4846       NaN       NaN       NaN       NaN       5.0  
4847       NaN       NaN       NaN       NaN       5.0  

[5 rows x 207 columns]

In [2]:
data.shape

(4848, 207)

In [3]:
data.columns

Index(['user_id', 'Movie1', 'Movie2', 'Movie3', 'Movie4', 'Movie5', 'Movie6',
       'Movie7', 'Movie8', 'Movie9',
       ...
       'Movie197', 'Movie198', 'Movie199', 'Movie200', 'Movie201', 'Movie202',
       'Movie203', 'Movie204', 'Movie205', 'Movie206'],
      dtype='object', length=207)

In [4]:
data.isnull().sum()

user_id        0
Movie1      4847
Movie2      4847
Movie3      4847
Movie4      4846
            ... 
Movie202    4842
Movie203    4847
Movie204    4840
Movie205    4813
Movie206    4835
Length: 207, dtype: int64

In [5]:
data.max()
#data[data['user_id']=="AZZ1KF8RAO1BR"]

user_id     AZZ1KF8RAO1BR
Movie1                  5
Movie2                  5
Movie3                  2
Movie4                  5
                ...      
Movie202                5
Movie203                3
Movie204                5
Movie205                5
Movie206                5
Length: 207, dtype: object

In [6]:
data_main=data.copy()

In [7]:
data.describe()

Movie1  Movie2  Movie3  Movie4     Movie5  Movie6  Movie7  Movie8  \
count     1.0     1.0     1.0     2.0  29.000000     1.0     1.0     1.0   
mean      5.0     5.0     2.0     5.0   4.103448     4.0     5.0     5.0   
std       NaN     NaN     NaN     0.0   1.496301     NaN     NaN     NaN   
min       5.0     5.0     2.0     5.0   1.000000     4.0     5.0     5.0   
25%       5.0     5.0     2.0     5.0   4.000000     4.0     5.0     5.0   
50%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
75%       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   
max       5.0     5.0     2.0     5.0   5.000000     4.0     5.0     5.0   

       Movie9  Movie10  ...  Movie197  Movie198  Movie199  Movie200  Movie201  \
count     1.0      1.0  ...  5.000000       2.0       1.0  8.000000  3.000000   
mean      5.0      5.0  ...  3.800000       5.0       5.0  4.625000  4.333333   
std       NaN      NaN  ...  1.643168       0.0       NaN  0.517549  1.154701   
min       5.0      5.0  ...  1.000000       5.0       5.0  4.000000  3.000000   
25%       5.0      5.0  ...  4.000000       5.0       5.0  4.000000  4.000000   
50%       5.0      5.0  ...  4.000000       5.0       5.0  5.000000  5.000000   
75%       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   
max       5.0      5.0  ...  5.000000       5.0       5.0  5.000000  5.000000   

       Movie202  Movie203  Movie204   Movie205   Movie206  
count  6.000000       1.0  8.000000  35.000000  13.000000  
mean   4.333333       3.0  4.375000   4.628571   4.923077  
std    1.632993       NaN  1.407886   0.910259   0.277350  
min    1.000000       3.0  1.000000   1.000000   4.000000  
25%    5.000000       3.0  4.750000   5.000000   5.000000  
50%    5.000000       3.0  5.000000   5.000000   5.000000  
75%    5.000000       3.0  5.000000   5.000000   5.000000  
max    5.000000       3.0  5.000000   5.000000   5.000000  

[8 rows x 206 columns]

In [8]:
# Transposing data frame df
data.describe().T["count"].sort_values(ascending=False)[:10].to_frame()

count
Movie127  2313.0
Movie140   578.0
Movie16    320.0
Movie103   272.0
Movie29    243.0
Movie91    128.0
Movie92    101.0
Movie89     83.0
Movie158    66.0
Movie108    54.0

# 1> Which movies have maximum views/ratings? 
Ans: Movie127.
We can see from above Movie1 has rating=5, movie2 has 5, movie3 has 2 and so on.
Movie127 has rated 2313 times. 

In [9]:
 data.drop('user_id',axis=1).sum().sort_values(ascending=False)[:10].to_frame()
#data.drop('user_id',axis=1).mean().sort_values(ascending=False)[:10].to_frame()

0
Movie127  9511.0
Movie140  2794.0
Movie16   1446.0
Movie103  1241.0
Movie29   1168.0
Movie91    586.0
Movie92    482.0
Movie89    380.0
Movie158   318.0
Movie108   252.0

# 2> What is the average rating for each movie? Define the top 5 movies with the maximum ratings.
From above table, we can see that 
average ratings for Movie127 is 9511/2313 =4.11
average ratings for Movie140 is 2794/578 =4.83
average ratings for Movie16 is 1446/320 =4.51
average ratings for Movie103 is 1241/272 =4.56
average ratings for Movie29 is 1168/243 =4.80

In [10]:
data.describe().T["count"].sort_values(ascending=True)[:5].to_frame()

count
Movie1      1.0
Movie71     1.0
Movie145    1.0
Movie69     1.0
Movie68     1.0

# 3> Define the top 5 movies with the least audience.
Ans:-Above 5 Movies has been rated only once.

# Recommendation model:

4> Divide the data into training and test data
5> Build a recommendation model on training data
6> Make predictions on the test data

We have to fill those Nan values with our prediction.

In [ ]:
# !pip install scikit-surprise 
# Anaconda prompt >conda install -c conda-forge scikit-surprise

In [11]:
from surprise import Reader
from surprise import accuracy
from surprise.model_selection import train_test_split

In [25]:
# help(data.melt())

In [13]:
data_main.columns

Index(['user_id', 'Movie1', 'Movie2', 'Movie3', 'Movie4', 'Movie5', 'Movie6',
       'Movie7', 'Movie8', 'Movie9',
       ...
       'Movie197', 'Movie198', 'Movie199', 'Movie200', 'Movie201', 'Movie202',
       'Movie203', 'Movie204', 'Movie205', 'Movie206'],
      dtype='object', length=207)

In [14]:
melt_data=data_main.melt(id_vars=data_main.columns[0],value_vars=data_main.columns[1:],var_name="movie_name",value_name="rating")
melt_data

user_id movie_name  rating
0       A3R5OBKS7OM2IR     Movie1     5.0
1        AH3QC2PC1VTGP     Movie1     NaN
2       A3LKP6WPMP9UKX     Movie1     NaN
3        AVIY68KEPQ5ZD     Movie1     NaN
4       A1CV1WROP5KTTW     Movie1     NaN
...                ...        ...     ...
998683  A1IMQ9WMFYKWH5   Movie206     5.0
998684  A1KLIKPUF5E88I   Movie206     5.0
998685   A5HG6WFZLO10D   Movie206     5.0
998686  A3UU690TWXCG1X   Movie206     5.0
998687   AI4J762YI6S06   Movie206     5.0

[998688 rows x 3 columns]

In [15]:
from surprise import Dataset
reader=Reader(rating_scale=(-1,10))
data=Dataset.load_from_df(melt_data.fillna(0), reader=reader)
trainset,testset=train_test_split(data,test_size=0.25)
from surprise import SVD
SHalgo=SVD()
SHalgo.fit(trainset)

In [16]:
predictions=SHalgo.test(testset)
accuracy.rmse(predictions)
# this rmse is not good

RMSE: 0.2803


0.2803474511234512

In [17]:
user_id='A3R5OBKS7OM2IR'
muvi_id='Movie1'
r_ui=5.0
SHalgo.predict(user_id,muvi_id,r_ui=r_ui,verbose=True)
# we can see that est = 0.08  so estimated value is not close to 5.00. 

user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 0.08   {'was_impossible': False}


Prediction(uid='A3R5OBKS7OM2IR', iid='Movie1', r_ui=5.0, est=0.07706739906089126, details={'was_impossible': False})

In [18]:
# Instead of filling value to zero, we will fill it with mean and median and then will check rmse value again.
# applying cross validation

from surprise.model_selection import cross_validate
cross_validate(SHalgo,data,measures=['RMSE','MAE'],cv=3,verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.2861  0.2819  0.2763  0.2814  0.0040  
MAE (testset)     0.0436  0.0424  0.0425  0.0428  0.0005  
Fit time          31.75   32.00   32.43   32.06   0.28    
Test time         2.25    2.05    2.13    2.15    0.08    


{'test_rmse': array([0.28613262, 0.28186272, 0.27631906]),
 'test_mae': array([0.04357285, 0.04244687, 0.04252051]),
 'fit_time': (31.74809193611145, 31.999433755874634, 32.433354139328),
 'test_time': (2.252018928527832, 2.0534684658050537, 2.1342875957489014)}

In [19]:
def Func_CV(algo_type,frame,min_,max_):
    reader=Reader(rating_scale=(min_,max_))
    data=Dataset.load_from_df(frame, reader=reader)
    algo=algo_type
    cross_validate(algo,data,measures=['RMSE','MAE'],cv=3,verbose=True)
    user_id='A3R5OBKS7OM2IR'
    muvi_id='Movie1'
    r_ui=5.0
    print(algo.predict(user_id,muvi_id,r_ui=r_ui,verbose=True))
    
# Made our recommendations to a one particular user

In [20]:
data_main=data_main.iloc[:1500,:50] # taking first 1500 rows and 50 columns
melt_data=data_main.melt(id_vars=data_main.columns[0],value_vars=data_main.columns[1:],var_name="movie_name",value_name="rating")
Func_CV(SVD(),melt_data.fillna(0),-1,10)
Func_CV(SVD(),melt_data.fillna(melt_data.mean()),-1,10)
Func_CV(SVD(),melt_data.fillna(melt_data.median()),-1,10)
# i have reduced my dataset- only for one id
# mean is working fine here so I will test mean on full dataset. RMSE is .08 (it has significantly reduced from fillna(0))
# and est is 4.63 which is close to r_ui=5.
# I am not taking median because median might make model overfitted.

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.4175  0.4404  0.4006  0.4195  0.0163  
MAE (testset)     0.0898  0.0914  0.0854  0.0889  0.0026  
Fit time          2.18    2.22    2.22    2.21    0.02    
Test time         0.27    0.12    0.11    0.17    0.07    
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 0.47   {'was_impossible': False}
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 0.47   {'was_impossible': False}
Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    0.0804  0.0769  0.0856  0.0810  0.0036  
MAE (testset)     0.0158  0.0158  0.0155  0.0157  0.0001  
Fit time          2.22    2.19    2.20    2.20    0.01    
Test time         0.11    0.27    0.11    0.16    0.07    
user: A3R5OBKS7OM2IR item: Movie1     r_ui = 5.00   est = 4.63   {'was_impossible': False}
user: A3R5OBKS7OM2IR item: Mo

In [21]:
# gridsearch
from surprise.model_selection  import GridSearchCV
param_grid= {'n_epochs':[20,30],
             # in 1st iteration it will do gradient descent 20 times
            'lr_all':[0.005,0.01], 
             # for 1st iteration my step distance=.005
            'n_factors':[50,100]} 
             # for 1st iteration it will check 50 features

In [26]:
SHgs=GridSearchCV(SVD,param_grid, measures=['rmse','mse'],cv=3) 
# instead of cv=3, u can pass list
SHgs.fit(data)

In [23]:
SHgs.best_score

{'rmse': 0.2777008871654881, 'mse': 0.0771284095064293}

In [24]:
print(SHgs.best_params['rmse']) 
# printing combination of parameters that gave best RMSE score.

{'n_epochs': 30, 'lr_all': 0.01, 'n_factors': 50}
